R/estimateEvenness.R

```diff
@@ -0,0 +1,177 @@
+#'
+#' @param threshold a numeric threshold. assay values below or equal to this
+#'   threshold will be set to zero.
+#'
+#' @param BPPARAM A
+#'   \code{\link[BiocParallel:BiocParallelParam-class]{BiocParallelParam}}
+#'   object specifying whether calculation of estimates should be parallelized.
+#'
+#' @param ... additional parameters passed to \code{estimateEvenness}
+#'
+#' @return \code{x} with additional \code{\link{colData}} named
+#'   \code{*name*}
+#'
+#' @seealso
+#' \code{\link[scater:plotColData]{plotColData}}
+#'
+#' @name estimateEvenness
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' se <- GlobalPatterns
+#'
+#' se <- estimateEvenness(se)
+#' head(colData(se))
+NULL
+
+#' @rdname estimateEvenness
+#' @export
+setGeneric("estimateEvenness",signature = c("x"),
+           function(x, abund_values = "counts",
+                    index = c("camargo", "pielou", "simpson", "evar", "bulla"),
+                    name = index, ...)
+               standardGeneric("estimateEvenness"))
+
+#' @rdname estimateEvenness
+#' @export
+setMethod("estimateEvenness", signature = c(x = "SummarizedExperiment"),
+    function(x, abund_values = "counts",
+             index = c("camargo", "pielou", "simpson", "evar", "bulla"),
+             name = index, ..., BPPARAM = SerialParam()){
+        # input check
+        index<- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                 "same length than 'index'.",
+                 call. = FALSE)
+        }
+        .check_abund_values(abund_values, x)
+        #
+        vnss <- BiocParallel::bplapply(index,
+                                       .run_evenness,
+                                       mat = assay(x, abund_values),
+                                       BPPARAM = BPPARAM, ...)
+        .add_values_to_colData(x, vnss, name)
+    }
+)
+
+.calc_bulla_evenness <- function(mat) {
+    # Species richness (number of species)
+    S <- colSums2(mat > 0, na.rm = TRUE)
+
+    # Relative abundances
+    p <- t(mat)/colSums2(mat, na.rm = TRUE)
+
+    i <- seq_len(nrow(p))
+    O <- vapply(i,function(i){sum(pmin(p[i,], 1/S[i]))},numeric(1))
+
+    # Bulla's Evenness
+    (O - 1/S)/(1 - 1/S)
+}
+
+# Camargo's eveness x: species counts zeroes: include zeros Inspired
+# by code from Pepijn de Vries and Zhou Xiang at
+# researchgate.net/post/How_can_we_calculate_the_Camargo_evenness_index_in_R
+# but rewritten here
+.calc_camargo_evenness <- function(mat) {
+    N <- colSums2(mat > 0, na.rm = TRUE)
+
+    seq <- IntegerList(lapply(N - 1,seq_len))
+
+    x <- mapply(
+        function(i, n, s){
+            xx <- 0
+            for (j in s) {
+                xx <- xx + sum(abs(mat[(j + 1):n,i] - mat[j,i]))
+            }
+            xx
+        },
+        seq_along(N),
+        N,
+        seq)
+    # Return
+    1 - x/(colSums2(mat, na.rm = TRUE) * N)
```

Comment:
same comment with colSums2(mat, na.rm=TRUE) - used multiple times in the different evenness calculators.

Reply:
Wrapping these single functional calls, wouldn't have any performance effects and maintainability would also not be better. Arguments to function would need to be fix and the interdepences between the indicies would make it hard to reason about.

Reply:
Performance is fine already, so this is a minor issue. But I do not see how it could not lead to performance improvement as the same operation gets calculated now multiple times in different functions. Instead of calculating this 4-5 times it would be possible to do just once and feed into the functions. I agree that this does not necessarily improve maintainability.

Reply:
I wouldn't touch it.

Any optimisation in this regard, would have to be done by the introduction of if clauses and I really don't want to catch and arrange all the cases needed to be covered